docs/src/rfcs/00X-global-buffer.md

```diff
@@ -0,0 +1,324 @@
+  
+Similar to Global, GlobalMut will also be introduced:
+
+      #[allow(unused_attributes)]
+      #[spirv(global)]
+      pub struct GlobalMut<'a, T> {
+          x: &'a mut T
+      }
+
+      impl<'a, T> GlobalMut<'a, T> {
+          pub(crate) fn new(x: &'a mut T) -> Self {
+              Self { x }
+          }
+      }
+
+      impl<T: Copy> GlobalMut<'a, T> {
+          pub fn load(&self) -> T {
+              *self.x
+          }
+          pub fn store(&mut self, x: T) {
+              *self.x = T;
+          }
+      }
+
+      impl<'a, T> GlobalMut<'a, [T]> {
+          pub fn get<I: SliceIndex<[T]>>(&self, index: I) -> Option<Global<'a, <I as SliceIndex<[T]>>::Output>> {
+              self.x.get(index)
+                  .map(|x| Global::new(x))
+          } 
+          pub(crate) unsafe fn get_unchecked<I: SliceIndex<[T]>(&self, index: I) -> Global<'a, <I as SliceIndex<[T]>>::Output {
+              Global::new(self.x.get_unchecked(index))
+          }
+          pub fn get_mut<I: SliceIndexMut<[T]>>(&mut self, index: I) -> Option<GlobalMut<'a, <I as SliceIndexMut<[T]>>::Output>> {
+              self.x.get(index)
+                  .map(|x| GlobalMut::new(x))
+          } 
+          pub(crate) unsafe fn get_unchecked_mut<I: SliceIndex<[T]>(&mut self, index: I) -> GlobalMut<'a, <I as SliceIndexMut<[T]>>::Output> {
+              GlobalMut::new(self.x.get_unchecked(index))
+          }
+      }
+
+The intent with Global and GlobalMut, is that they are returned from higher level abstractions. SPIR-V has a requirement that with several pointer operations, including Load, Store, and AccessChain, the storage class of the pointer matches the storage class of the variable. That means if we have a buffer, ie a *{StorageBuffer} RuntimeArray, we can't return a plain reference or plain slice, ie *{Function} T, to the user to perform the derefence. There are potentially alternatives, but the simplest approach is to just maintain the wrapped Global / GlobalMut struct until the store or load. This is the current design of spirv-std. 
+
+Global and GlobalMut are essentially either &(mut)T or &(mut)\[T\]. They are only to be provided to user code when they are not aliased. Safe abstractions should ensure that this is the case, and then yield these objects. 
+
+Due to the nature of GlobalMut, it is critical that the compiler prevents GlobalMut from being passed in as a parameter, since it marks a safe to access reference or slice. Global, since it is immutable, is not problematic, but might as well be banned as well, as it isn't meant to be used this way, and will not be portable. Further, the spirv attribute, at least for "storage_buffer" and "cross_workgroup", etc, should be limited only to within the crate. 
```

Comment:
I'm not sure I follow exactly why `GlobalMut` shouldn't be passed to functions, would you mind elaborating this a bit?

Reply:
Not functions, but as a parameter to entry points. All invocations share GlobalBuffer, but the intent is that Global is then a safe, per invocation reference. If the user can just declare a GlobalMut as a parameter, it can freely be accessed without protections against multiple mutation that GlobalBufferMut would provide.

Reply:
✅